crates/uv-resolver/src/candidate_selector.rs

```diff
@@ -290,9 +289,16 @@ impl CandidateSelector {
                 let allow = match self.prerelease_strategy.allows(package_name, env) {
                     AllowPrerelease::Yes => true,
                     AllowPrerelease::No => false,
-                    // If the pre-release is "global" (i.e., provided via a lockfile, rather than
-                    // a fork), accept it unless pre-releases are completely banned.
-                    AllowPrerelease::IfNecessary => marker.is_true(),
+                    // If the pre-release is "global", i.e., it comes from a solve without a fork,
+                    // or is provided via a lockfile, accept it unless pre-releases are completely
+                    // banned.
+                    AllowPrerelease::IfNecessary => match source {
+                        PreferenceSource::Fork => false,
+                        PreferenceSource::Lock
+                        | PreferenceSource::Environment
+                        | PreferenceSource::Resolve
+                        | PreferenceSource::RequirementsTxt => true,
+                    },
```

Comment:
This is the crux of the change. I believe relying on `marker.is_true()` to determine if we should respect a preference was fallible for cases where the preference came an existing fork in the lockfile with markers attatched.

Reply:
I'm down for trying it.

Reply:
The PR suggests we have scenarios to test this already, so worth trying I think.